air/src/trace/rows.rs

```diff
@@ -0,0 +1,349 @@
+        write!(f, "{}", self.0)
+    }
+}
+
+// FROM ROW INDEX
+// ================================================================================================
+
+impl From<RowIndex> for u32 {
+    fn from(step: RowIndex) -> u32 {
+        step.0
+    }
+}
+
+impl From<RowIndex> for u64 {
+    fn from(step: RowIndex) -> u64 {
+        step.0 as u64
+    }
+}
+
+impl From<RowIndex> for usize {
+    fn from(step: RowIndex) -> usize {
+        step.0 as usize
+    }
+}
+
+impl From<RowIndex> for Felt {
+    fn from(step: RowIndex) -> Felt {
+        Felt::from(step.0)
+    }
+}
+
+// INTO ROW INDEX
+// ================================================================================================
+
+/// Converts a usize value into a [`RowIndex`].
+///
+/// # Panics
+///
+/// This function will panic if the number represented by the usize is greater than the maximum
+/// [`RowIndex`] value, `u32::MAX`.
+impl From<usize> for RowIndex {
+    fn from(value: usize) -> Self {
+        let value = u32::try_from(value).map_err(|_| RowIndexError::InvalidSize(value)).unwrap();
+        value.into()
+    }
+}
+
+/// Converts a u64 value into a [`RowIndex`].
+///
+/// # Panics
+///
+/// This function will panic if the number represented by the u64 is greater than the maximum
+/// [`RowIndex`] value, `u32::MAX`.
+impl TryFrom<u64> for RowIndex {
+    type Error = RowIndexError<u64>;
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        let value = u32::try_from(value).map_err(|_| RowIndexError::InvalidSize(value))?;
+        Ok(RowIndex::from(value))
+    }
+}
+
+impl From<u32> for RowIndex {
+    fn from(value: u32) -> Self {
+        Self(value)
+    }
+}
+
+/// Converts an i32 value into a [`RowIndex`].
+///
+/// # Panics
+///
+/// This function will panic if the number represented by the i32 is less than 0.
+impl From<i32> for RowIndex {
+    fn from(value: i32) -> Self {
+        let value = u32::try_from(value).map_err(|_| RowIndexError::InvalidSize(value)).unwrap();
+        RowIndex(value)
+    }
+}
+
+// ROW INDEX OPS
+// ================================================================================================
+
+/// Subtracts a usize from a [`RowIndex`].
+///
+/// # Panics
+///
+/// This function will panic if the number represented by the usize is greater than the maximum
+/// [`RowIndex`] value, `u32::MAX`.
+impl Sub<usize> for RowIndex {
+    type Output = RowIndex;
+
+    fn sub(self, rhs: usize) -> Self::Output {
+        let rhs = u32::try_from(rhs).map_err(|_| RowIndexError::InvalidSize(rhs)).unwrap();
+        RowIndex(self.0 - rhs)
+    }
+}
+
+impl Sub<RowIndex> for usize {
+    type Output = RowIndex;
+
+    fn sub(self, rhs: RowIndex) -> Self::Output {
+        (self - rhs.0 as usize).into()
+    }
+}
+
+impl Sub<RowIndex> for RowIndex {
+    type Output = RowIndex;
+
+    fn sub(self, rhs: RowIndex) -> Self::Output {
+        RowIndex(self.0 - rhs.0)
+    }
+}
```

Comment:
Similar to the above comment: what are some situations when we need to subtract row index from row index? (conceptually, this shouldn't be happening).

Reply:
```
                // difference between current clock cycle and start clock cycle of the current
                // asmop
                u32::from(a.max(b) - a.min(b.into())) as u8
```

```
                    // compute delta as difference between context IDs, addresses, or clock cycles
                    let delta = if prev_ctx != ctx {
                        (u32::from(ctx) - u32::from(prev_ctx)).into()
                    } else if prev_addr != addr {
                        u64::from(addr - prev_addr)
                    } else {
                        clk - prev_clk - 1
                    };
```

```
impl ChipletsLengths {
    pub fn new(chiplets: &Chiplets) -> Self {
        ChipletsLengths {
            hash_chiplet_len: chiplets.bitwise_start().into(),
            bitwise_chiplet_len: (chiplets.memory_start() - chiplets.bitwise_start()).into(),
            memory_chiplet_len: (chiplets.kernel_rom_start() - chiplets.memory_start()).into(),
            kernel_rom_len: (chiplets.padding_start() - chiplets.kernel_rom_start()).into(),
        }
    }
```

These are all of them.

Reply:
if these are all of them, I'd probably remove these operator overloading and do manual conversions here.